File: storage/data_original/at_coder/arc097/B/4803745.py
class UnionFind:
    def __init__(self, n):
        self.par = [i for i in range(n)]
        self.rank = [0]*n
    
    def find(self, x):
        if self.par[x] == x:
            return x
        else:
            self.par[x] = self.find(self.par[x])
            return self.par[x]
    
    def union(self, x, y):
        x = self.find(x)
        y = self.find(y)
        
        if self.rank[x] < self.rank[y]:
            self.par[x] = y
        else:
            self.par[y] = x
        
        if self.rank[x] == self.rank[y]:
            self.rank[x] += 1
    
    def same_check(self, x, y):
        return self.find(x) == self.find(y)


N, M = map(int, input().split())
p = [0] + [int(_) for _ in input().split()]
pair = [[int(_) for _ in input().split()] for _ in range(M)]
uf = UnionFind(N+1)
ans = 0

for i in pair:
    uf.union(i[0], i[1])

for i in range(1, N+1):
    if uf.find(i) == uf.find(p[i]):
        ans += 1

print(ans)

File: data/data_transpiled/at_coder/arc097/B/4803745.java
public static void UnionFind ( ) {
  class UnionFind {
    int [ ] par = new int [ n ] ;
    int [ ] rank = new int [ n ] ;
    int x ;
    int y ;
    public int find ( int x ) {
      if ( par [ x ] == x ) {
        return x ;
      }
      else {
        par [ x ] = find ( par [ x ] ) ;
        return x ;
      }
    }
    void union ( int x , int y ) {
      x = find ( x ) ;
      y = find ( y ) ;
      if ( rank [ x ] < rank [ y ] ) {
        par [ x ] = y ;
      }
      else {
        par [ y ] = x ;
      }
      if ( rank [ x ] == rank [ y ] ) {
        rank [ x ] ++ ;
      }
    }
    void sameCheck ( int x , int y ) {
    }
    int N = Integer . parseInt ( input ( ) ) ;
    int M = Integer . parseInt ( input ( ) ) ;
    int [ ] p = new int [ M ] ;
    for ( int i = 0 ;
    i < p . length ;
    i ++ ) {
      p [ i ] = Integer . parseInt ( input ( ) ) ;
    }
    int [ ] [ ] pair = new int [ M ] [ M ] ;
    for ( int i = 0 ;
    i < pair . length ;
    i ++ ) {
      pair [ i ] = pair [ i ] [ 0 ] ;
    }
    UnionFind uf = new UnionFind ( N + 1 ) ;
    int ans = 0 ;
    for ( int i = 0 ;
    i < pair . length ;
    i ++ ) {
      uf . union ( pair [ i ] [ 0 ] , pair [ i ] [ 1 ] ) ;
    }
    for ( int i = 1 ;
    i <= N ;
    i ++ ) {
      if ( uf . find ( i ) == uf . find ( p [ i ] ) ) ans ++ ;
    }
    System . out . println ( ans ) ;
  }
  